File: scripts/deepagents_sdk_smoke.py
#!/usr/bin/env python3
"""No-cost smoke test of the pinned Deep Agents SDK and bounded tool surface."""

from __future__ import annotations

import argparse
import importlib.util
import json
import os
import shlex
import socket
import tempfile
from datetime import datetime, timezone
from importlib.metadata import version
from pathlib import Path
from typing import Any
from unittest import mock

PACKAGE_ROOT = Path(__file__).resolve().parents[1]
WORKER_PATH = PACKAGE_ROOT / "scripts" / "deepagents_worker.py"
ARTIFACTS = PACKAGE_ROOT / "artifacts"


def _load_worker() -> Any:
    spec = importlib.util.spec_from_file_location("deepagents_worker", WORKER_PATH)
    if spec is None or spec.loader is None:
        raise RuntimeError("cannot load Deep Agents worker")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def _scripted_model(
    *, forbidden_tool: str | None = None, command_target: Path | None = None
) -> Any:
    """Create the fake model lazily so importing this script needs no SDK extras."""
    from collections.abc import Callable, Iterator, Sequence  # noqa: PLC0415

    from langchain_core.language_models import LanguageModelInput  # noqa: PLC0415
    from langchain_core.language_models.chat_models import BaseChatModel  # noqa: PLC0415
    from langchain_core.messages import AIMessage, BaseMessage  # noqa: PLC0415
    from langchain_core.outputs import ChatGeneration, ChatResult  # noqa: PLC0415
    from langchain_core.runnables import Runnable  # noqa: PLC0415
    from langchain_core.tools import BaseTool  # noqa: PLC0415
    from pydantic import Field  # noqa: PLC0415

    class ScriptedChatModel(BaseChatModel):
        messages: Iterator[AIMessage] = Field(exclude=True)
        observed_tools: list[str] = Field(default_factory=list)

        @property
        def _llm_type(self) -> str:
            return "openai"

        def _get_ls_params(self, **_: Any) -> dict[str, str]:
            return {"ls_provider": "openai", "ls_model_name": "gpt-5.2-codex"}

        def bind_tools(
            self,
            tools: Sequence[dict[str, Any] | type | Callable | BaseTool],
            **_: Any,
        ) -> Runnable[LanguageModelInput, AIMessage]:
            self.observed_tools = [
                tool.name
                if isinstance(tool, BaseTool)
                else str(tool.get("function", {}).get("name", ""))
                if isinstance(tool, dict)
                else getattr(tool, "__name__", str(tool))
                for tool in tools
            ]
            return self

        def _generate(
            self,
            messages: list[BaseMessage],
            stop: list[str] | None = None,
            run_manager: Any = None,
            **kwargs: Any,
        ) -> ChatResult:
            del messages, stop, run_manager, kwargs
            return ChatResult(generations=[ChatGeneration(message=next(self.messages))])

    if forbidden_tool is not None:
        forbidden_arguments: dict[str, dict[str, Any]] = {
            "delete": {"file_path": "/app/delete-canary.txt"},
            "execute": {
                "command": f"touch {shlex.quote(str(command_target))}"
                if command_target is not None
                else "exit 0"
            },
            "task": {"description": "forbidden synthetic subagent dispatch"},
            "write_todos": {"todos": []},
        }
        if forbidden_tool not in forbidden_arguments:
            raise ValueError(f"unsupported forbidden tool probe: {forbidden_tool}")
        messages = [
            AIMessage(
                content="",
                tool_calls=[
                    {
                        "name": forbidden_tool,
                        "args": forbidden_arguments[forbidden_tool],
                        "id": f"forbidden-{forbidden_tool}-dispatch",
                        "type": "tool_call",
                    }
                ],
            ),
            AIMessage(content="A forbidden dispatch unexpectedly returned."),
        ]
    else:
        messages = [
            AIMessage(
                content="",
                tool_calls=[
                    {
                        "name": "write_file",
                        "args": {
                            "file_path": "/tests/forbidden.txt",
                            "content": "must-not-exist\n",
                        },
                        "id": "write-forbidden-smoke",
                        "type": "tool_call",
                    }
                ],
            ),
            AIMessage(
                content="",
                tool_calls=[
                    {
                        "name": "write_file",
                        "args": {
                            "file_path": "/../escape-write.txt",
                            "content": "must-not-escape\n",
                        },
                        "id": "write-traversal-smoke",
                        "type": "tool_call",
                    }
                ],
            ),
            AIMessage(
                content="",
                tool_calls=[
                    {
                        "name": "read_file",
                        "args": {"file_path": "/../outside-secret.txt"},
                        "id": "read-traversal-smoke",
                        "type": "tool_call",
                    }
                ],
            ),
            AIMessage(
                content="",
                tool_calls=[
                    {
                        "name": "read_file",
                        "args": {"file_path": "/app/value.txt"},
                        "id": "read-smoke",
                        "type": "tool_call",
                    }
                ],
            ),
            AIMessage(
                content="",
                tool_calls=[
                    {
                        "name": "edit_file",
                        "args": {
                            "file_path": "/app/value.txt",
                            "old_string": "before\n",
                            "new_string": "after\n",
                        },
                        "id": "edit-smoke",
                        "type": "tool_call",
                    }
                ],
            ),
            AIMessage(content="Candidate edit prepared; controller verification required."),
        ]
    scripted = ScriptedChatModel(messages=iter(messages))
    return scripted


def _error_contains(error: BaseException, expected: str) -> bool:
    pending: list[BaseException] = [error]
    seen: set[int] = set()
    while pending:
        current = pending.pop()
        if id(current) in seen:
            continue
        seen.add(id(current))
        if expected in str(current):
            return True
        nested = getattr(current, "exceptions", ())
        if isinstance(nested, (list, tuple)):
            pending.extend(item for item in nested if isinstance(item, BaseException))
        for item in (current.__cause__, current.__context__):
            if isinstance(item, BaseException):
                pending.append(item)
    return False


def _assert_forbidden_dispatch_rejected(
    *, worker: Any, workspace: Path, packet: dict[str, Any]
) -> bool:
    delete_canary = workspace / "app" / "delete-canary.txt"
    execute_canary = workspace / "app" / "execute-canary.txt"
    delete_canary.write_text("must-remain\n", encoding="utf-8")
    for tool_name in ("execute", "delete", "task", "write_todos"):
        model = _scripted_model(forbidden_tool=tool_name, command_target=execute_canary)
        agent = worker.build_bounded_agent(
            model=model,
            profile_key="openai",
            workspace=workspace,
            packet=packet,
        )
        try:
            agent.invoke(
                {"messages": [{"role": "user", "content": "Run the synthetic probe."}]},
                config={"recursion_limit": 8},
            )
        except BaseException as exc:
            if not _error_contains(exc, f"forbidden tool call: {tool_name}"):
                raise AssertionError(
                    f"{tool_name} was not rejected by the controller dispatch boundary"
                ) from exc
        else:
            raise AssertionError(f"forbidden tool call unexpectedly completed: {tool_name}")
    if execute_canary.exists() or delete_canary.read_text(encoding="utf-8") != "must-remain\n":
        raise AssertionError("a forbidden tool dispatch changed the workspace")
    return True


def _smoke_passed(
    *,
    workspace_edit_succeeded: bool,
    out_of_scope_write_denied: bool,
    traversal_write_denied: bool,
    traversal_read_denied: bool,
    tool_names: list[str],
    expected_tools: list[str],
    network_attempts: list[str],
    forbidden_tool_calls_rejected: bool,
) -> bool:
    return (
        workspace_edit_succeeded
        and out_of_scope_write_denied
        and traversal_write_denied
        and traversal_read_denied
        and tool_names == expected_tools
        and {"delete", "execute", "task", "write_todos"}.isdisjoint(tool_names)
        and forbidden_tool_calls_rejected
        and not network_attempts
    )


def _assert_os_network_disabled() -> None:
    errors: list[str] = []
    try:
        socket.getaddrinfo("docs.langchain.com", 443)
    except OSError:
        pass
    else:
        errors.append("DNS resolution unexpectedly succeeded")
    probe = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    probe.settimeout(1)
    try:
        probe.connect(("1.1.1.1", 443))
    except OSError:
        pass
    else:
        errors.append("external socket connection unexpectedly succeeded")
    finally:
        probe.close()
    if errors:
        raise RuntimeError("; ".join(errors))


def run_smoke(
    *, output: Path | None = None, assert_os_network_disabled: bool = False
) -> dict[str, Any]:
    if assert_os_network_disabled:
        _assert_os_network_disabled()
    packet = {
        "schema_version": 1,
        "policy": {
            "controller_is_sole_acceptor": True,
            "allowed_paths": ["app/"],
        },
    }
    network_attempts: list[str] = []

    def deny_network(*args: Any, **kwargs: Any) -> Any:
        del kwargs
        network_attempts.append(repr(args[:2]))
        raise RuntimeError("network access is disabled during the SDK smoke")

    disabled_environment = {
        "LANGSMITH_TRACING": "false",
        "LANGCHAIN_TRACING_V2": "false",
        "DEEPAGENTS_CODE_OFFLINE": "1",
        "DEEPAGENTS_CODE_NO_UPDATE_CHECK": "1",
    }
    with (
        tempfile.TemporaryDirectory(prefix="daiw-sdk-smoke-") as temporary,
        mock.patch.dict(os.environ, disabled_environment, clear=False),
        mock.patch("socket.create_connection", side_effect=deny_network),
        mock.patch("socket.getaddrinfo", side_effect=deny_network),
        mock.patch.object(socket.socket, "connect", side_effect=deny_network),
    ):
        worker = _load_worker()
        actual_version = version("deepagents")
        if actual_version != worker.EXPECTED_DEEPAGENTS_VERSION:
            raise RuntimeError(
                f"deepagents=={worker.EXPECTED_DEEPAGENTS_VERSION} is required; "
                f"found {actual_version}"
            )
        scripted = _scripted_model()
        workspace = Path(temporary) / "workspace"
        target = workspace / "app" / "value.txt"
        forbidden_target = workspace / "tests" / "forbidden.txt"
        escaped_write_target = Path(temporary) / "escape-write.txt"
        outside_secret = Path(temporary) / "outside-secret.txt"
        outside_canary = "DAIW_OUTSIDE_READ_CANARY"
        target.parent.mkdir(parents=True)
        target.write_text("before\n", encoding="utf-8")
        outside_secret.write_text(outside_canary + "\n", encoding="utf-8")
        forbidden_tool_calls_rejected = _assert_forbidden_dispatch_rejected(
            worker=worker,
            workspace=workspace,
            packet=packet,
        )
        agent = worker.build_bounded_agent(
            model=scripted,
            profile_key="openai",
            workspace=workspace,
            packet=packet,
        )
        result = agent.invoke(
            {"messages": [{"role": "user", "content": "Apply the scripted edit."}]},
            config={"recursion_limit": 18},
        )
        final_content = worker._final_content(result)
        tool_results = {
            getattr(message, "tool_call_id", None): str(getattr(message, "content", ""))
            for message in result.get("messages", [])
            if getattr(message, "tool_call_id", None)
        }
        traversal_write_output = tool_results.get("write-traversal-smoke", "").lower()
        traversal_read_output = tool_results.get("read-traversal-smoke", "")
        traversal_write_denied = not escaped_write_target.exists() and any(
            marker in traversal_write_output for marker in ("denied", "not allowed", "error")
        )
        traversal_read_denied = outside_canary not in traversal_read_output and any(
            marker in traversal_read_output.lower() for marker in ("denied", "not allowed", "error")
        )
        tool_names = sorted(scripted.observed_tools)
        expected_tools = sorted(worker.ALLOWED_TOOLS)
        workspace_edit_succeeded = target.read_text(encoding="utf-8") == "after\n"
        out_of_scope_write_denied = not forbidden_target.exists()
        passed = _smoke_passed(
            workspace_edit_succeeded=workspace_edit_succeeded,
            out_of_scope_write_denied=out_of_scope_write_denied,
            traversal_write_denied=traversal_write_denied,
            traversal_read_denied=traversal_read_denied,
            tool_names=tool_names,
            expected_tools=expected_tools,
            network_attempts=network_attempts,
            forbidden_tool_calls_rejected=forbidden_tool_calls_rejected,
        )
        record = {
            "schema_version": 1,
            "at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
            "runtime": "deepagents",
            "runtime_version": actual_version,
            "network_request_made": bool(network_attempts),
            "network_attempts": len(network_attempts),
            "model": "openai:gpt-5.2-codex (scripted, no transport)",
            "profile_provider": "openai",
            "observed_tools": tool_names,
            "forbidden_tools_absent": all(
                item not in tool_names for item in ("delete", "execute", "task", "write_todos")
            ),
            "forbidden_tool_calls_rejected": forbidden_tool_calls_rejected,
            "workspace_edit_succeeded": workspace_edit_succeeded,
            "out_of_scope_write_denied": out_of_scope_write_denied,
            "traversal_write_denied": traversal_write_denied,
            "traversal_read_denied": traversal_read_denied,
            "final_response_present": bool(final_content),
            "passed": passed,
        }
    output = output or ARTIFACTS / "deepagents-sdk-smoke.json"
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text(json.dumps(record, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    return {"artifact": str(output), **record}


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--output", type=Path)
    parser.add_argument("--assert-os-network-disabled", action="store_true")
    args = parser.parse_args()
    result = run_smoke(
        output=args.output,
        assert_os_network_disabled=args.assert_os_network_disabled,
    )
    print(json.dumps(result, indent=2, sort_keys=True))
    return 0 if result["passed"] else 1


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript-runtime/src/deepagents_sdk_smoke.ts
#!/usr/bin/env node
/** No-cost smoke of the pinned TypeScript SDK and exact bounded tool surface. */

import dns from "node:dns";
import dnsPromises from "node:dns/promises";
import * as fsPromises from "node:fs/promises";
import { syncBuiltinESMExports } from "node:module";
import * as net from "node:net";
import * as os from "node:os";
import * as path from "node:path";
import { pathToFileURL } from "node:url";

type JsonObject = Record<string, unknown>;
const EXPECTED_NODE_VERSION = "22.23.2";

function isObject(value: unknown): value is JsonObject {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function contentText(value: unknown): string {
  if (typeof value === "string") return value;
  return JSON.stringify(value);
}

function toolResults(result: unknown): Map<string, string> {
  const values = new Map<string, string>();
  if (!isObject(result) || !Array.isArray(result.messages)) return values;
  for (const message of result.messages) {
    if (!isObject(message)) continue;
    const rawId = message.tool_call_id ?? message.toolCallId;
    if (typeof rawId === "string") {
      values.set(rawId, contentText(message.content));
    }
  }
  return values;
}

function finalResponsePresent(result: unknown): boolean {
  if (!isObject(result) || !Array.isArray(result.messages) || result.messages.length === 0) {
    return false;
  }
  const last = result.messages[result.messages.length - 1];
  return isObject(last) && contentText(last.content).length > 0;
}

const DNS_METHODS = [
  "lookup",
  "lookupService",
  "resolve",
  "resolve4",
  "resolve6",
  "resolveAny",
  "resolveCaa",
  "resolveCname",
  "resolveMx",
  "resolveNaptr",
  "resolveNs",
  "resolvePtr",
  "resolveSoa",
  "resolveSrv",
  "resolveTxt",
  "reverse",
] as const;

function installNetworkDeny(attempts: string[]): () => void {
  const originalFetch = globalThis.fetch;
  const originalConnect = net.Socket.prototype.connect;
  const restores: Array<() => void> = [];
  const deny = (kind: string): never => {
    attempts.push(kind);
    throw new Error("network access is disabled during the TypeScript SDK smoke");
  };
  const patchMethods = (target: object, prefix: string): void => {
    const mutable = target as Record<string, unknown>;
    for (const method of DNS_METHODS) {
      const original = mutable[method];
      if (typeof original !== "function") continue;
      mutable[method] = (..._args: unknown[]) => deny(`${prefix}.${method}`);
      restores.push(() => {
        mutable[method] = original;
      });
    }
  };
  globalThis.fetch = ((..._args: Parameters<typeof fetch>) => deny("fetch")) as typeof fetch;
  net.Socket.prototype.connect = function (..._args: unknown[]): net.Socket {
    return deny("socket.connect");
  } as typeof net.Socket.prototype.connect;
  patchMethods(dns, "dns");
  patchMethods(dns.Resolver.prototype, "dns.Resolver");
  patchMethods(dnsPromises, "dns.promises");
  patchMethods(dnsPromises.Resolver.prototype, "dns.promises.Resolver");
  syncBuiltinESMExports();
  return () => {
    for (const restore of restores.reverse()) restore();
    globalThis.fetch = originalFetch;
    net.Socket.prototype.connect = originalConnect;
    syncBuiltinESMExports();
  };
}

async function assertOsNetworkDisabled(): Promise<void> {
  const failures: string[] = [];
  try {
    await dnsPromises.lookup("docs.langchain.com");
    failures.push("DNS resolution unexpectedly succeeded");
  } catch {
    // Expected under the container's network-none boundary.
  }
  await new Promise<void>((resolve) => {
    const socket = net.createConnection({ host: "1.1.1.1", port: 443 });
    socket.setTimeout(1000);
    socket.once("connect", () => {
      failures.push("external socket connection unexpectedly succeeded");
      socket.destroy();
      resolve();
    });
    socket.once("error", () => resolve());
    socket.once("timeout", () => {
      socket.destroy();
      resolve();
    });
  });
  if (failures.length > 0) throw new Error(failures.join("; "));
}

async function scriptedModel() {
  const [{ AIMessage }, { fakeModel }] = await Promise.all([
    import("@langchain/core/messages"),
    import("@langchain/core/testing"),
  ]);
  const scripted = fakeModel();
  const calls = [
    {
      name: "write_file",
      id: "write-forbidden-smoke",
      args: { file_path: "/tests/forbidden.txt", content: "must-not-exist\n" },
    },
    {
      name: "write_file",
      id: "write-traversal-smoke",
      args: { file_path: "/../escape-write.txt", content: "must-not-escape\n" },
    },
    {
      name: "read_file",
      id: "read-traversal-smoke",
      args: { file_path: "/../outside-secret.txt" },
    },
    {
      name: "read_file",
      id: "read-smoke",
      args: { file_path: "/app/value.txt" },
    },
    {
      name: "edit_file",
      id: "edit-smoke",
      args: {
        file_path: "/app/value.txt",
        old_string: "before\n",
        new_string: "after\n",
      },
    },
  ];
  for (const call of calls) scripted.respondWithTools([call]);
  scripted.respond(new AIMessage("Candidate edit prepared; controller verification required."));
  Object.defineProperty(scripted, "getName", {
    configurable: false,
    value: () => "ChatOpenAI",
    writable: false,
  });
  return scripted;
}

export async function runSmoke(options: {
  output?: string;
  assertOsNetworkDisabled?: boolean;
} = {}): Promise<JsonObject> {
  if (process.versions.node !== EXPECTED_NODE_VERSION) {
    throw new Error(`Node ${EXPECTED_NODE_VERSION} is required; found ${process.versions.node}`);
  }
  if (options.assertOsNetworkDisabled) await assertOsNetworkDisabled();
  const temporary = await fsPromises.mkdtemp(path.join(os.tmpdir(), "daiw-typescript-sdk-smoke-"));
  const workspace = path.join(temporary, "workspace");
  const target = path.join(workspace, "app", "value.txt");
  const forbiddenTarget = path.join(workspace, "tests", "forbidden.txt");
  const escapedTarget = path.join(temporary, "escape-write.txt");
  const outsideSecret = path.join(temporary, "outside-secret.txt");
  const outsideCanary = "DAIW_TYPESCRIPT_OUTSIDE_READ_CANARY";
  const networkAttempts: string[] = [];
  const restoreNetwork = installNetworkDeny(networkAttempts);
  try {
    // The guard is deliberately active before either the worker or SDK module
    // graph is evaluated. Any import-time fetch, socket, or DNS request fails
    // closed and is counted in this controller-validated smoke record.
    const worker = await import("./deepagents_worker.js");
    if (worker.EXPECTED_NODE_VERSION !== EXPECTED_NODE_VERSION) {
      throw new Error("worker and smoke Node version policy diverged");
    }
    await fsPromises.mkdir(path.dirname(target), { recursive: true });
    await fsPromises.writeFile(target, "before\n", "utf8");
    await fsPromises.writeFile(outsideSecret, `${outsideCanary}\n`, "utf8");
    const model = await scriptedModel();
    const bounded = await worker.buildBoundedAgent({
      model,
      profileKey: "openai",
      workspace,
      packet: {
        schema_version: 1,
        policy: { controller_is_sole_acceptor: true, allowed_paths: ["app/"] },
      },
    });
    const result = await bounded.agent.invoke(
      { messages: [{ role: "user", content: "Apply the scripted edit." }] },
      { recursionLimit: 24 },
    );
    const results = toolResults(result);
    const traversalWriteOutput = (results.get("write-traversal-smoke") ?? "").toLowerCase();
    const traversalReadOutput = results.get("read-traversal-smoke") ?? "";
    const exactTools = bounded.observedTools();
    const forbiddenAbsent = worker.FORBIDDEN_TOOLS.every((name) => !exactTools.includes(name));
    const workspaceEditSucceeded = (await fsPromises.readFile(target, "utf8")) === "after\n";
    const outOfScopeWriteDenied = !(await fsPromises
      .lstat(forbiddenTarget)
      .then(() => true)
      .catch(() => false));
    const traversalWriteDenied =
      !(await fsPromises
        .lstat(escapedTarget)
        .then(() => true)
        .catch(() => false)) &&
      ["denied", "not allowed", "error", "outside"].some((marker) =>
        traversalWriteOutput.includes(marker),
      );
    const traversalReadDenied =
      !traversalReadOutput.includes(outsideCanary) &&
      ["denied", "not allowed", "error", "outside"].some((marker) =>
        traversalReadOutput.toLowerCase().includes(marker),
      );
    const hasFinalResponse = finalResponsePresent(result);
    const passed =
      workspaceEditSucceeded &&
      outOfScopeWriteDenied &&
      traversalWriteDenied &&
      traversalReadDenied &&
      JSON.stringify(exactTools) === JSON.stringify([...worker.ALLOWED_TOOLS].sort()) &&
      forbiddenAbsent &&
      hasFinalResponse &&
      networkAttempts.length === 0;
    const record: JsonObject = {
      schema_version: 1,
      at: new Date().toISOString(),
      runtime: "deepagents",
      sdk_language: "typescript",
      runtime_version: worker.EXPECTED_DEEPAGENTS_VERSION,
      node_version: process.versions.node,
      model: "openai:scripted-smoke (scripted, no transport)",
      profile_provider: "openai",
      model_transport: "scripted-no-transport",
      network_request_made: networkAttempts.length > 0,
      network_attempts: networkAttempts.length,
      observed_tools: exactTools,
      forbidden_tools_absent: forbiddenAbsent,
      workspace_edit_succeeded: workspaceEditSucceeded,
      out_of_scope_write_denied: outOfScopeWriteDenied,
      traversal_write_denied: traversalWriteDenied,
      traversal_read_denied: traversalReadDenied,
      final_response_present: hasFinalResponse,
      passed,
    };
    if (options.output) {
      await fsPromises.writeFile(options.output, `${JSON.stringify(record, null, 2)}\n`, {
        encoding: "utf8",
        flag: "wx",
        // The network-none container runs as UID 65532 while the host-side
        // controller validates this synthetic record under its own UID. The
        // parent temporary directory remains private; the record must be
        // host-readable across that UID boundary.
        mode: 0o644,
      });
    }
    return record;
  } finally {
    restoreNetwork();
    await fsPromises.rm(temporary, { recursive: true, force: true });
  }
}

function parseArguments(argv: string[]): { output?: string; assertOsNetworkDisabled: boolean } {
  let output: string | undefined;
  let assertNetwork = false;
  for (let index = 0; index < argv.length; index += 1) {
    const value = argv[index];
    if (value === "--assert-os-network-disabled") {
      assertNetwork = true;
    } else if (value === "--output" && index + 1 < argv.length) {
      output = argv[index + 1];
      index += 1;
    } else {
      throw new Error(`unsupported argument: ${String(value)}`);
    }
  }
  return output === undefined
    ? { assertOsNetworkDisabled: assertNetwork }
    : { output, assertOsNetworkDisabled: assertNetwork };
}

const invokedPath = process.argv[1] ? pathToFileURL(path.resolve(process.argv[1])).href : "";
if (import.meta.url === invokedPath) {
  runSmoke(parseArguments(process.argv.slice(2)))
    .then((record) => {
      process.stdout.write(`${JSON.stringify(record, null, 2)}\n`);
      if (record.passed !== true) process.exitCode = 1;
    })
    .catch((error: unknown) => {
      process.stderr.write(`${error instanceof Error ? error.message : "smoke failed"}\n`);
      process.exitCode = 1;
    });
}
